README.rst
A Python structured logger for Fluentd
======================================

WARNING: This is a fork of the https://github.com/fluent/fluent-logger-python
project to work with asyncio.


Many web/mobile applications generate huge amount of event logs (c,f.
login, logout, purchase, follow, etc). To analyze these event logs could
be really valuable for improving the service. However, the challenge is
collecting these logs easily and reliably.

`Fluentd <https://github.com/fluent/fluentd>`__ solves that problem by
having: easy installation, small footprint, plugins, reliable buffering,
log forwarding, etc.

**aiofluent** is a Python library, to record the events from
Python application.

Requirements
------------

-  Python 3.5 or greater
- ``msgpack-python``

Installation
------------

This library is distributed as 'aiofluent' python package. Please
execute the following command to install it.

.. code:: sh

    $ pip install aiofluent

Configuration
-------------

Fluentd daemon must be launched with a tcp source configuration:

::

    <source>
      type forward
      port 24224
    </source>

To quickly test your setup, add a matcher that logs to the stdout:

::

    <match app.**>
      type stdout
    </match>

Usage
-----

FluentSender Interface
~~~~~~~~~~~~~~~~~~~~~~

`sender.FluentSender` is a structured event logger for Fluentd.

By default, the logger assumes fluentd daemon is launched locally. You
can also specify remote logger by passing the options.

.. code:: python

    from aiofluent import sender

    # for local fluent
    logger = sender.FluentSender('app')

    # for remote fluent
    logger = sender.FluentSender('app', host='host', port=24224)

For sending event, call `emit` method with your event. Following example will send the event to
fluentd, with tag 'app.follow' and the attributes 'from' and 'to'.

.. code:: python

    # Use current time
    logger.emit('follow', {'from': 'userA', 'to': 'userB'})

    # Specify optional time
    cur_time = int(time.time())
    logger.emit_with_time('follow', cur_time, {'from': 'userA', 'to':'userB'})

You can detect an error via return value of `emit`. If an error happens in `emit`, `emit` returns `False` and get an error object using `last_error` method.

.. code:: python

    if not logger.emit('follow', {'from': 'userA', 'to': 'userB'}):
        print(logger.last_error)
        logger.clear_last_error() # clear stored error after handled errors

If you want to shutdown the client, call `close()` method.

.. code:: python

    logger.close()

Event-Based Interface
~~~~~~~~~~~~~~~~~~~~~

This API is a wrapper for `sender.FluentSender`.

First, you need to call ``sender.setup()`` to create global `sender.FluentSender` logger
instance. This call needs to be called only once, at the beginning of
the application for example.

Initialization code of Event-Based API is below:

.. code:: python

    from aiofluent import sender

    # for local fluent
    sender.setup('app')

    # for remote fluent
    sender.setup('app', host='host', port=24224)

Then, please create the events like this. This will send the event to
fluentd, with tag 'app.follow' and the attributes 'from' and 'to'.

.. code:: python

    from aiofluent import event

    # send event to fluentd, with 'app.follow' tag
    event.Event('follow', {
      'from': 'userA',
      'to':   'userB'
    })

`event.Event` has one limitation which can't return success/failure result.

Other methods for Event-Based Interface.

.. code:: python

    sender.get_global_sender # get instance of global sender
    sender.close # Call FluentSender#close

Handler for buffer overflow
~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can inject your own custom proc to handle buffer overflow in the event of connection failure. This will mitigate the loss of data instead of simply throwing data away.

.. code:: python

    import msgpack
    from io import BytesIO

    def handler(pendings):
        unpacker = msgpack.Unpacker(BytesIO(pendings))
        for unpacked in unpacker:
            print(unpacked)

    logger = sender.FluentSender('app', host='host', port=24224, buffer_overflow_handler=handler)

You should handle any exception in handler. aiofluent ignores exceptions from ``buffer_overflow_handler``.

This handler is also called when pending events exist during `close()`.

Python logging.Handler interface
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This client-library also has ``FluentHandler`` class for Python logging
module.

.. code:: python

    import logging
    from aiofluent import handler

    custom_format = {
      'host': '%(hostname)s',
      'where': '%(module)s.%(funcName)s',
      'type': '%(levelname)s',
      'stack_trace': '%(exc_text)s'
    }

    logging.basicConfig(level=logging.INFO)
    l = logging.getLogger('fluent.test')
    h = handler.FluentHandler('app.follow', host='host', port=24224)
    formatter = handler.FluentRecordFormatter(custom_format)
    h.setFormatter(formatter)
    l.addHandler(h)
    l.info({
      'from': 'userA',
      'to': 'userB'
    })
    l.info('{"from": "userC", "to": "userD"}')
    l.info("This log entry will be logged with the additional key: 'message'.")

You can also customize formatter via logging.config.dictConfig

.. code:: python

    import logging.config
    import yaml

    with open('logging.yaml') as fd:
        conf = yaml.load(fd)

    logging.config.dictConfig(conf['logging'])

A sample configuration ``logging.yaml`` would be:

.. code:: python

    logging:
        version: 1

        formatters:
          brief:
            format: '%(message)s'
          default:
            format: '%(asctime)s %(levelname)-8s %(name)-15s %(message)s'
            datefmt: '%Y-%m-%d %H:%M:%S'
          fluent_fmt:
            '()': fluent.handler.FluentRecordFormatter
            format:
              level: '%(levelname)s'
              hostname: '%(hostname)s'
              where: '%(module)s.%(funcName)s'

        handlers:
            console:
                class : logging.StreamHandler
                level: DEBUG
                formatter: default
                stream: ext://sys.stdout
            fluent:
                class: fluent.handler.FluentHandler
                host: localhost
                port: 24224
                tag: test.logging
                formatter: fluent_fmt
                level: DEBUG
            none:
                class: logging.NullHandler

        loggers:
            amqp:
                handlers: [none]
                propagate: False
            conf:
                handlers: [none]
                propagate: False
            '': # root logger
                handlers: [console, fluent]
                level: DEBUG
                propagate: False

License
-------

Apache License, Version 2.0


CHANGELOG.rst
1.2.10 (unreleased)
-------------------

- Nothing changed yet.


1.2.9 (2020-10-22)
------------------

- Only log errors every 30 seconds


1.2.8 (2020-05-15)
------------------

- Handle TypeError formatting log data


1.2.7 (2020-03-09)
------------------

- Fix repo location


1.2.6 (2020-01-06)
------------------

- Improve error logging
  [vangheem]

1.2.5 (2019-12-19)
------------------

- Handle event loop closed error
  [vangheem]


1.2.4 (2019-12-19)
------------------

- Increase max queue size


1.2.3 (2019-04-01)
------------------

- Fix release


1.2.2 (2019-04-01)
------------------

- nanosecond_precision by default
  [davidonna]

1.2.1 (2018-10-31)
------------------

- Add support for nanosecond precision timestamps
  [davidonna]

1.2.0 (2018-06-14)
------------------

- Maintain one AsyncIO queue for all logs
  [vangheem]

1.1.4 (2018-05-29)
------------------

- Handle RuntimeError on canceling tasks/cleanup
  [vangheem]


1.1.3 (2018-02-15)
------------------

- Lock calling the close method of sender
  [vangheem]

- Increase default timeout
  [vangheem]


1.1.2 (2018-02-07)
------------------

- lock the whole method
  [vangheem]


1.1.1 (2018-02-07)
------------------

- Use lock on getting connection object
  [vangheem]


1.1.0 (2018-01-25)
------------------

- Move to using asyncio connection infrastructure instead of sockets
  [vangheem]


1.0.8 (2018-01-04)
------------------

- Always close out buffer data
  [vangheem]


1.0.7 (2018-01-04)
------------------

- Handle errors processing log queue
  [vangheem]


1.0.6 (2017-11-14)
------------------

- Prevent log queue from getting too large
  [vangheem]


1.0.5 (2017-10-17)
------------------

- Fix release to include CHANGELOG.rst file
  [vangheem]


1.0.4 (2017-10-10)
------------------

- Fix pushing initial record


1.0.3 (2017-10-10)
------------------

- Handle Runtime error when logging done before event loop started
  [vangheem]


1.0.2 (2017-10-09)
------------------

- Fix to make normal logging call async
  [vangheem]


1.0.1 (2017-07-03)
------------------

- initial release


tests/fixtures.py
from aiofluent import sender
from tests import mockserver

import pytest


@pytest.fixture(scope="function")
async def mock_server():
    server = mockserver.MockRecvServer()
    yield server


@pytest.fixture(scope="function")
def test_sender(mock_server):
    sender.setup('app', connection_factory=mock_server.factory)
    yield sender
    sender.close()
    sender._set_global_sender(None)


@pytest.fixture(scope="function")
def mock_sender(mock_server):
    msender = sender.FluentSender(
        tag='test', connection_factory=mock_server.factory)
    yield msender
    msender.close()


tests/mockserver.py
# -*- coding: utf-8 -*-
from io import BytesIO
from msgpack import Unpacker


class Writer:

    def __init__(self, server):
        self.server = server

    def write(self, data):
        self.server._buf.write(data)

    async def drain(self):
        pass

    def close(self):
        pass


class MockRecvServer:

    def __init__(self):
        self._writer = Writer(self)
        self._buf = BytesIO()

    async def factory(self, sender):
        return None, self._writer

    def get_recieved(self):
        self._buf.seek(0)
        return list(Unpacker(self._buf, encoding='utf-8'))


tests/test_event.py
# -*- coding: utf-8 -*-

from aiofluent import event, sender

import pytest


@pytest.mark.asyncio
async def test_logging(test_sender):
    # XXX: This tests succeeds even if the fluentd connection failed
    # send event with tag app.follow
    await event.send_event('follow', {
        'from': 'userA',
        'to': 'userB'
    })


@pytest.mark.asyncio
async def test_logging_with_timestamp(test_sender):
    # send event with tag app.follow, with timestamp
    await event.send_event('follow', {
        'from': 'userA',
        'to': 'userB'
    }, time=int(0))


@pytest.mark.asyncio
async def test_no_last_error_on_successful_event(test_sender):
    global_sender = sender.get_global_sender()
    await event.send_event('unfollow', {
        'from': 'userC',
        'to': 'userD'
    })

    assert global_sender.last_error is None
    sender.close()


aiofluent/sender.py
# -*- coding: utf-8 -*-
import asyncio
import msgpack
import socket
import sys
import time
import traceback
import struct

_global_sender = None


def _set_global_sender(sender):
    """ [For testing] Function to set global sender directly
    """
    global _global_sender
    _global_sender = sender


def setup(tag, **kwargs):
    global _global_sender
    _global_sender = FluentSender(tag, **kwargs)


def get_global_sender():
    return _global_sender


def close():
    get_global_sender().close()


async def connection_factory(sender):
    try:
        return await asyncio.wait_for(
            asyncio.open_connection(sender._host, sender._port),
            sender._timeout)
    except (asyncio.TimeoutError, asyncio.CancelledError) as ex:
        sys.stderr.write('Timeout connecting to fluentd')
        sender.last_error = ex
    except Exception as ex:
        sys.stderr.write('Unknown error connecting to fluentd')
        sender.last_error = ex


class EventTime(msgpack.ExtType):
    def __new__(cls, timestamp):
        seconds = int(timestamp)
        nanoseconds = int(timestamp % 1 * 10 ** 9)
        return super(EventTime, cls).__new__(
            cls,
            code=0,
            data=struct.pack(">II", seconds, nanoseconds),
        )


class FluentSender(object):
    def __init__(self,
                 tag,
                 host='localhost',
                 port=24224,
                 bufmax=1 * 1024 * 1024,
                 timeout=3,
                 verbose=False,
                 buffer_overflow_handler=None,
                 retry_timeout=30,
                 connection_factory=connection_factory,
                 nanosecond_precision=True,
                 **kwargs):

        self._tag = tag
        self._host = host
        self._port = port
        self._bufmax = bufmax
        self._timeout = timeout
        self._verbose = verbose
        self._buffer_overflow_handler = buffer_overflow_handler
        self._nanosecond_precision = nanosecond_precision

        self._pendings = None
        self._reader = None
        self._writer = None
        self._retry_timeout = retry_timeout

        self._last_error = None
        self._last_error_time = 0
        self._lock = None

        self._connection_factory = connection_factory

    @property
    def lock(self):
        if self._lock is None:
            self._lock = asyncio.Lock()
        return self._lock

    async def get_writer(self):
        async with self.lock:
            if self._writer is not None:
                return self._writer

            if (self._last_error_time + self._retry_timeout) > time.time():
                return

            result = await self._connection_factory(self)
            if result:
                self._reader, self._writer = result
            return self._writer

    async def async_emit(self, label, data, timestamp=None):
        if timestamp is not None:
            ev_time = timestamp
        elif self._nanosecond_precision:
            ev_time = EventTime(time.time())
        else:
            ev_time = int(time.time())
        return await self.async_emit_with_time(label, ev_time, data)

    async def async_emit_with_time(self, label, timestamp, data):
        if self._nanosecond_precision and isinstance(timestamp, float):
            timestamp = EventTime(timestamp)
        try:
            bytes_ = self._make_packet(label, timestamp, data)
        except Exception as e:
            self.last_error = e
            bytes_ = self._make_packet(label, timestamp,
                                       {"level": "CRITICAL",
                                        "message": "Can't output to log",
                                        "traceback": traceback.format_exc()})
        return await self._async_send(bytes_)

    def _make_packet(self, label, timestamp, data):
        if label:
            tag = '.'.join((self._tag, label))
        else:
            tag = self._tag
        packet = (tag, timestamp, data)
        if self._verbose:
            print(packet)
        return msgpack.packb(packet)

    async def _async_send(self, bytes_):
        try:
            result = await self._async_send_internal(bytes_)
        except Exception:
            result = None
        return result

    async def _async_send_internal(self, bytes_):
        # buffering
        if self._pendings:
            self._pendings += bytes_
            bytes_ = self._pendings

        try:
            writer = await self.get_writer()
            if writer is None:
                self.clean(bytes_)
                return False
            writer.write(bytes_)
            await asyncio.wait_for(writer.drain(), self._timeout)

            self._pendings = None
            self._last_error_time = 0
            return True
        except (socket.error, asyncio.TimeoutError,
                asyncio.CancelledError, OSError, BlockingIOError) as e:
            self.last_error = e

            # Connection error, retry connecting
            self.clean(bytes_)
            async with self.lock:
                self.close()
            self._writer = self._reader = None
            return False
        except Exception as ex:
            self.last_error = ex
            sys.stderr.write('Unhandled exception sending data')
            self.clean(bytes_)
            return False

    def clean(self, bytes_=b''):
        if self._pendings and (len(self._pendings) > self._bufmax):
            self._call_buffer_overflow_handler(self._pendings)
            self._pendings = None
        else:
            self._pendings = bytes_

    def _call_buffer_overflow_handler(self, pending_events):
        try:
            if self._buffer_overflow_handler:
                self._buffer_overflow_handler(pending_events)
        except Exception:
            # User should care any exception in handler
            pass

    @property
    def last_error(self):
        return self._last_error

    @last_error.setter
    def last_error(self, err):
        if err is not None:
            self._last_error = err
            self._last_error_time = time.time()
        else:
            self.clear_last_error()

    def clear_last_error(self):
        self._last_error = None
        self._last_error_time = 0

    def close(self):
        if self._writer is not None:
            try:
                self._writer.close()
            except RuntimeError:
                # event loop already closed
                pass
            self._reader = None
            self._writer = None


aiofluent/handler.py
# -*- coding: utf-8 -*-

from aiofluent import sender

import asyncio
import json
import logging
import socket
import sys
import time
import traceback


class FluentRecordFormatter(logging.Formatter, object):
    """ A structured formatter for Fluent.

    Best used with server storing data in an ElasticSearch cluster for example.

    :param fmt: a dict with format string as values to map to provided keys.
    """
    def __init__(self, fmt=None, datefmt=None, style='%'):
        super(FluentRecordFormatter, self).__init__(None, datefmt)

        if not fmt:
            self._fmt_dict = {
                'sys_host': '%(hostname)s',
                'sys_name': '%(name)s',
                'sys_module': '%(module)s',
            }
        else:
            self._fmt_dict = fmt

        self.hostname = socket.gethostname()

    def format(self, record):
        # Compute attributes handled by parent class.
        super(FluentRecordFormatter, self).format(record)
        # Add ours
        record.hostname = self.hostname
        # Apply format
        data = {}
        for key, value in self._fmt_dict.items():
            try:
                data[key] = value % record.__dict__
            except (KeyError, TypeError):
                # we are okay with missing values here...
                pass

        self._structuring(data, record)
        return data

    def usesTime(self):
        return any([value.find('%(asctime)') >= 0
                    for value in self._fmt_dict.values()])

    def _structuring(self, data, record):
        """ Melds `msg` into `data`.

        :param data: dictionary to be sent to fluent server
        :param msg: :class:`LogRecord`'s message to add to `data`.
          `msg` can be a simple string for backward compatibility with
          :mod:`logging` framework, a JSON encoded string or a dictionary
          that will be merged into dictionary generated in :meth:`format.
        """
        msg = record.msg

        if isinstance(msg, dict):
            self._add_dic(data, msg)
        elif isinstance(msg, str):
            try:
                self._add_dic(data, json.loads(str(msg)))
            except ValueError:
                msg = record.getMessage()
                self._add_dic(data, {'message': msg})
        else:
            self._add_dic(data, {'message': msg})

    @staticmethod
    def _add_dic(data, dic):
        for key, value in dic.items():
            if isinstance(key, str):
                data[str(key)] = value


MAX_QUEUE_SIZE = 500


class LogQueue:

    def __init__(self, queue=None):
        self._queue = queue

    async def consume_queue(self, initial_record, handler):
        if self._queue is None:
            self._queue = asyncio.Queue(maxsize=MAX_QUEUE_SIZE)
        self._queue.put_nowait((initial_record, handler, time.time()))
        while True:
            record, handler, timestamp = await self._queue.get()
            try:
                await handler.async_emit(record, timestamp)
            except:  # noqa
                sys.stderr.write(
                    'Error processing log\n{}\n'.format(
                        traceback.format_exc()
                    ))
            finally:
                self._queue.task_done()

    def qsize(self):
        if self._queue is None:
            return 0
        return self._queue.qsize()

    def put_nowait(self, *args):
        self._queue.put_nowait(*args)


class FluentHandler(logging.Handler):
    '''
    Logging Handler for fluent.
    '''

    # class singletons
    _queue = None
    _queue_task = None

    def __init__(self,
                 tag,
                 host='localhost',
                 port=24224,
                 timeout=3,
                 verbose=False,
                 loop=None,
                 nanosecond_precision=False,
                 **kwargs):
        self.loop = loop
        self.tag = tag
        self.nanosecond_precision = nanosecond_precision
        self.sender = sender.FluentSender(tag,
                                          host=host, port=port,
                                          timeout=timeout, verbose=verbose,
                                          nanosecond_precision=self.nanosecond_precision,
                                          **kwargs)
        self.last_warning_sent = 0
        logging.Handler.__init__(self)

    def emit(self, record):
        if FluentHandler._queue_task is None or FluentHandler._queue_task.done():
            # the queue should be a singleton, we don't need a task
            # for every log handler
            try:
                FluentHandler._queue = LogQueue()
                FluentHandler._queue_task = asyncio.ensure_future(
                    FluentHandler._queue.consume_queue(record, self), loop=self.loop)
            except RuntimeError:
                sys.stderr.write(
                    'No event loop running to send log to fluentd\n')
        else:
            try:
                FluentHandler._queue.put_nowait((record, self, time.time()))
            except RuntimeError:
                sys.stderr.write("RuntimeError, likely event loop closing\n")
            except asyncio.QueueFull:
                if time.time() - self.last_warning_sent > 30:
                    sys.stderr.write(
                        f'Fluentd hit max log queue size({MAX_QUEUE_SIZE}), '
                        'discarding message\n')
                    self.last_warning_sent = time.time()
            except AttributeError:
                sys.stderr.write('Error sending async fluentd message\n')

    async def async_emit(self, record, timestamp=None):
        data = self.format(record)
        return await self.sender.async_emit(None, data, timestamp)

    def close(self):
        self.acquire()
        try:
            self.sender.close()
            logging.Handler.close(self)
            if self._queue_task is not None and not self._queue_task.done():
                try:
                    self._queue_task.cancel()
                except RuntimeError:
                    pass
        finally:
            self.release()
